packages/ember-glimmer/lib/helpers/each-in.js

```diff
@@ -0,0 +1,63 @@
+/**
+@module ember
+@submodule ember-templates
+*/
+
+import { UpdatableReference } from '../utils/references';
+import symbol from 'ember-metal/symbol';
+
+/**
+  The `{{each-in}}` helper loops over properties on an object. It is unbound,
+  in that new (or removed) properties added to the target object will not be
+  rendered.
+
+  For example, given a `user` object that looks like:
+
+  ```javascript
+  {
+    "name": "Shelly Sails",
+    "age": 42
+  }
+  ```
+
+  This template would display all properties on the `user`
+  object in a list:
+
+  ```handlebars
+  <ul>
+  {{#each-in user as |key value|}}
+    <li>{{key}}: {{value}}</li>
+  {{/each-in}}
+  </ul>
+  ```
+
+  Outputting their name and age.
+
+  @method each-in
+  @for Ember.Templates.helpers
+  @public
+  @since 2.1.0
+*/
+export default {
+  isInternalHelper: true,
+  toReference(args) {
+    return new EachInHelperReference(args.positional.at(0));
+  }
+};
+
+const EACH_IN_REFERENCE = symbol('EACH_IN');
+
+export function isEachIn(ref) {
+  return ref && ref[EACH_IN_REFERENCE];
+}
+
+class EachInHelperReference extends UpdatableReference {
+  constructor(sourceRef) {
+    super(sourceRef);
+    this[EACH_IN_REFERENCE] = true;
+  }
+
+  value() {
+    return super().value();
+  }
+}
```

Comment:
I think you can just do:

```javascript
class ProxyReference {
  constructor(inner) {
    this.tag = inner.tag;
    this.inner = inner;
  }

  value() {
    this.inner.value();
  }

  get(key) {
    return this.inner.get(key);
  }
}

class EachInReference extends ProxyReference {
  constructor(inner) {
    super(inner);
    this[EACH_IN_REFERENCE] = true;
  }
}
```

Might want to put the `ProxyReference` in utils.

One problem with this is that you loose the `const`-ness on the original reference and any extra things like `[UPDATE]`. For propagating const-ness, we can do this:

```javascript
class ProxyReference {
  static create(inner) {
    if (isConst(inner)) {
      return new ConstProxyReference(inner);
    } else {
      return new this(inner);
    }
  }

  // ...
}

class ConstProxyReference extends ConstReference {
  // ... repeat the same code
}
```

But now you have the `EachInReference` subclass problem :(

So at the end of the day, what we really want to do is:


```javascript
export default {
    isInternalHelper: true,
    toReference(args) {
      let wrapped = Object.create(args.positional.at(0));
      wrapped[EACH_IN_REFERENCE] = true;
      return wrapped;
    }
};
```

I know we usually want to avoid `Object.create`, but as you can see everything else have other composition problems, so maybe this is the best option (it only happens *once* per `{{#each-in}}` invocation during initial render anyway, so hopefully it is fine?

cc @krisselden